commit 2c56ad30da7c3746675f9730a56390d524e85117
Author: Christopher tofu Zorn <tofu@stanziq.com>
Date:   Thu Oct 29 10:40:20 2009 -0400

    anonymous sasl initializer, need to add it to the authenticator

diff --git a/twisted/words/protocols/jabber/sasl.py b/twisted/words/protocols/jabber/sasl.py
index 9a9b8e3..94435c7 100644
--- a/twisted/words/protocols/jabber/sasl.py
+++ b/twisted/words/protocols/jabber/sasl.py
@@ -126,13 +126,19 @@ class SASLInitiatingInitializer(xmlstream.BaseFeatureInitiatingInitializer):
         password = self.xmlstream.authenticator.password
 
         mechanisms = get_mechanisms(self.xmlstream)
-        if 'DIGEST-MD5' in mechanisms:
-            self.mechanism = sasl_mechanisms.DigestMD5('xmpp', jid.host, None,
-                                                       jid.user, password)
-        elif 'PLAIN' in mechanisms:
-            self.mechanism = sasl_mechanisms.Plain(None, jid.user, password)
+        if jid.user is not None:
+            if 'DIGEST-MD5' in mechanisms:
+                self.mechanism = sasl_mechanisms.DigestMD5('xmpp', jid.host, None,
+                                                           jid.user, password)
+            elif 'PLAIN' in mechanisms:
+                self.mechanism = sasl_mechanisms.Plain(None, jid.user, password)
+            else:
+                raise SASLNoAcceptableMechanism()
         else:
-            raise SASLNoAcceptableMechanism()
+            if 'ANONYMOUS' in mechanisms:
+                self.mechanism = sasl_mechanisms.Anonymous()
+            else:
+                raise sasl.SASLNoAccetableMechanisms()
 
     def start(self):
         """
diff --git a/twisted/words/protocols/jabber/sasl_mechanisms.py b/twisted/words/protocols/jabber/sasl_mechanisms.py
index 83d2280..6b84ef6 100644
--- a/twisted/words/protocols/jabber/sasl_mechanisms.py
+++ b/twisted/words/protocols/jabber/sasl_mechanisms.py
@@ -37,6 +37,17 @@ class ISASLMechanism(Interface):
 
 
 
+class Anonymous(object):
+    """Implements the ANONYMOUS SASL authentication mechanism.
+
+    This mechanism is defined in RFC 2245.
+    """
+    implements(ISASLMechanism)
+    name = 'ANONYMOUS'
+    
+    def getInitialResponse(self):
+        return None
+
 class Plain(object):
     """
     Implements the PLAIN SASL authentication mechanism.
diff --git a/twisted/words/test/test_jabbersaslmechanisms.py b/twisted/words/test/test_jabbersaslmechanisms.py
index 67b9876..f848579 100644
--- a/twisted/words/test/test_jabbersaslmechanisms.py
+++ b/twisted/words/test/test_jabbersaslmechanisms.py
@@ -17,6 +17,15 @@ class PlainTest(unittest.TestCase):
         m = sasl_mechanisms.Plain(None, 'test', 'secret')
         self.assertEquals(m.getInitialResponse(), '\x00test\x00secret')
 
+
+class AnonymousTest(unittest.TestCase):
+    def test_getInitialResponse(self):
+        """
+        Test the initial response.
+        """
+        m = sasl_mechanisms.Anonymous()
+        self.assertEquals(m.getInitialResponse(), None)
+
 class DigestMD5Test(unittest.TestCase):
     def setUp(self):
         self.mechanism = sasl_mechanisms.DigestMD5('xmpp', 'example.org', None,
